### PR TITLE
Customization of iOS audio categories from flutter.

### PR DESCRIPTION
**Pull request by chanonly123**
iOS specific only.
Now we can send this in `RecordConfig`, `iosAudioCategories`. 
```
const config = RecordConfig(encoder: encoder, numChannels: 1, iosConfig: IosRecordConfig(audioCategories: [IosAudioCategories.allowBluetooth]));
```
When recording starts, iOS recorder uses these categories.

### CODE REVIEW

File: record/example/lib/audio_recorder.dart [hunk 1]
```diff
@@ -53,7 +53,7 @@ class _RecorderState extends State<Recorder> with AudioRecorderMixin {
         final devs = await _audioRecorder.listInputDevices();
         debugPrint(devs.toString());
 
-        const config = RecordConfig(encoder: encoder, numChannels: 1);
+        const config = RecordConfig(encoder: encoder, numChannels: 1, iosAudioCategories: [IosAudioCategories.allowBluetoothA2DP]);
```
Comment (llfbandit) [inline record/example/lib/audio_recorder.dart:56]: Please let it with default values.

File: record_platform_interface/lib/src/types/record_config.dart [hunk 2]
```diff
@@ -51,6 +51,9 @@ class RecordConfig {
   /// Android specific configuration.
   final AndroidRecordConfig androidConfig;
 
+  /// iOS specific audioCategories
+  final List<IosAudioCategories> iosAudioCategories;
```
Comment (llfbandit) [inline record_platform_interface/lib/src/types/record_config.dart:55]: We should wrap those iOS specific categories in `iosConfig` of type `IosRecordConfig`.

File: record_platform_interface/lib/src/types/record_config.dart [hunk 3]
```diff
@@ -61,6 +64,7 @@ class RecordConfig {
     this.echoCancel = false,
     this.noiseSuppress = false,
     this.androidConfig = const AndroidRecordConfig(),
+    this.iosAudioCategories = const [IosAudioCategories.defaultToSpeaker, IosAudioCategories.allowBluetooth, IosAudioCategories.allowBluetoothA2DP],
```
Comment (llfbandit) [inline record_platform_interface/lib/src/types/record_config.dart:67]: 👍

File: record_darwin/ios/Classes/RecorderIOS.swift [hunk 4]
```diff
@@ -46,10 +46,31 @@ private func setInput(_ config: RecordConfig) throws {
...
+          if #available(iOS 9.0, *) { .interruptSpokenAudioAndMixWithOthers } else { nil }
+      case "allowBluetoothA2DP": 
+          if #available(iOS 10.0, *) { .allowBluetoothA2DP } else { nil }
+      case "allowAirPlay": 
+          if #available(iOS 10.0, *) { .allowAirPlay } else { nil }
+      case "overrideMutedMicrophoneInterruption": 
+          if #available(iOS 14.5, *) { .overrideMutedMicrophoneInterruption } else { nil }
+      default: nil
+      }
+    } ?? []
```
Comment (llfbandit) [inline record_darwin/ios/Classes/RecorderIOS.swift:70]: Should we re-apply previous default values?

**Review by llfbandit — CHANGES_REQUESTED** [review]

**Comment by llfbandit** [comment]
Thanks for your contribution!
Despite few changes requested, everything looks good! 👍

File: record_darwin/ios/Classes/RecorderIOS.swift (hunk 4 above)
Comment (llfbandit) [reply record_darwin/ios/Classes/RecorderIOS.swift:70]: No that's fine as categories will never be nil.

**Comment by chanonly123** [comment]
Mapping from `String` to `AVAudioSession.CategoryOptions` enums moved. Please check.

File: record_darwin/darwin/Classes/RecordConfig.swift [hunk 5]
```diff
@@ -37,27 +41,80 @@ public class RecordConfig {
...
+          .allowAirPlay
+      case "overrideMutedMicrophoneInterruption":
+        if #available(iOS 14.5, *) { .overrideMutedMicrophoneInterruption } else { nil }
+      default: nil
+      }
+    }
+    self.audioCategories = options ?? []
+  }
+
+  func toMap() -> [String: Any] {
```
Comment (llfbandit) [inline record_darwin/darwin/Classes/RecordConfig.swift:99]: `toMap` can be removed, it is never called. Or am I missing something?

**Review by llfbandit — APPROVED** [review]

**Comment by llfbandit** [comment]
Many thanks for first opening ios specific config!

**Comment by chanonly123** [comment]
Welcome. 
BTW tried to run macos project but `pod install` breaking. Can you help me here, then I can update for macos as well.
I am using `melos bootstrap` inside root.

**Comment by llfbandit** [comment]
In the example project?
What if you delete macos/Podfile.lock?

Also be sure to use a not too old version of cocoapod.

**Comment by tgalery** [comment]
I am also having some issues using the code incorporated in this MR:

I've added the following on my pubspec dependencies.

```
 record:
    git:
      url: https://github.com/llfbandit/record.git
      ref: master
      path: record
```

Looking at the pubspec.lock things seem to be updated to the `5.1.3` version and the `record_darwin` version to `1.1.2`. However the `record_platform_interface` seems to be unchanged and essentially this doesn't allow me to use the ios config.

Would it be possible to make a new release available to the general public with the changes ?

Thanks a lot !

**Comment by tgalery** [comment]
@chanonly123 could you specify the steps to use the code in this git repo to test locally ? My local env is resolving to hosted versions which do not contain the changes you introduce in this MR.

**Comment by chanonly123** [comment]
just run command `melos bootstrap`

**Pull request merged**